app/javascript/components/MapAndButtons.jsx

```diff
@@ -229,6 +233,10 @@ class MapAndButtons extends React.Component {
         </div>
         { this.render_modal() }
         <NotificationList errors={this.state.edit_locations.filter(l => l.hasError === true)} successes={this.state.success_notifications} success_notification_fade_delay={this.success_notification_fade_delay} />
+        <Legend persistent_locations={this.create_legend_set(this.props.persistent_locations)}
+                searched_locations={this.create_legend_set(this.props.locations)}
+                current_selected_floor={this.state.current_selected_floor.toString()}
```

Comment:
shouldn't need to pass the floor if you do the filtering in `create_legend_set`. Also.. `create_legend_set` should be set on the `state` and run in constructor, `componentDidUpdate` so that when the floor is changed, the legend gets new props passed to it.